### PR TITLE
[JENKINS-43094] Fingerprint snapshot dependencies

### PR DESCRIPTION
**Pull request by justinbarry**

### CODE REVIEW

File: maven-spy/src/main/java/org/jenkinsci/plugins/pipeline/maven/eventspy/handler/DependencyResolutionResultHandler.java [hunk 1]
```diff
@@ -42,13 +47,34 @@ protected boolean _handle(DependencyResolutionResult result) {
...
 
-        /*
         Xpp3Dom dependenciesElt = new Xpp3Dom("resolvedDependencies");
         root.addChild(dependenciesElt);
-        // don't cast to Dependency to not have a classloading issue at runtime
-        for (Object dependency: result.getResolvedDependencies()) {
-            dependenciesElt.addChild(newElement("dependency", dependency.toString()));
-        }*/
+
+        for (Dependency dependency: result.getResolvedDependencies()) {
```
Comment (cyrille-leclerc) [inline maven-spy/src/main/java/org/jenkinsci/plugins/pipeline/maven/eventspy/handler/DependencyResolutionResultHandler.java]: Didn't you get classloading issues? Did you try with Maven 3.5.0? I had issues with 3.3.9.

I saw when 3.5.0 was released that they did something with Aether but I did not investigate yet.

If it breaks with Maven <= 3.3.9 then we should have a proper exception handling recommending to use Maven 3.5.0 to benefit of the feature.

Comment (justinbarry) [inline maven-spy/src/main/java/org/jenkinsci/plugins/pipeline/maven/eventspy/handler/DependencyResolutionResultHandler.java]: I haven't seen an issue thus far. Our build slaves use either 3.2.5 or 3.3.9.

Is is possible another plugin is adding the dependency to the class path?

**Comment by justinbarry** [comment]
@cyrille-leclerc Thank for all the feedback. I hacked this together awhile back and haven't flush it out completely. Undoubtedly there are edge cases I haven't considered. Java isn't my first language so please forgive the copy and pasting. :)

File: jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/MavenSpyLogProcessor.java [hunk 2]
```diff
@@ -103,6 +99,14 @@ public void processMavenSpyLogs(StepContext context, FilePath mavenSpyLogFolder)
                     new GeneratedArtifactsReporter().process(context, mavenSpyLogsElt);
                 }
 
+                FilePath skipFingerprintDependenciesFile = workspace.child(".skip-archive-generated-artifacts");
+                if (skipFingerprintDependenciesFile.exists()) {
+                    listener.getLogger().println("[withMaven] Skip archiving of generated artifacts, file '" + skipFingerprintDependenciesFile + "' found in workspace");
```
Comment (justinbarry) [inline jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/MavenSpyLogProcessor.java]: Good idea.

**Comment by cyrille-leclerc** [comment]
> I hacked this together awhile back and haven't flush it out completely. Undoubtedly there are edge cases I haven't considered. Java isn't my first language so please forgive the copy and pasting. :)

No worries, I first need to solve [JENKINS-43596 Add withMaven(){} options for skipping reporters](https://issues.jenkins-ci.org/browse/JENKINS-43596) and then I'll work on your PR.

https://github.com/jenkinsci/pipeline-maven-plugin/pull/52

**Comment by cyrille-leclerc** [comment]
Replaced by #58

**Comment by cyrille-leclerc** [comment]
@justinbarry can you please test https://repo.jenkins-ci.org/releases/org/jenkins-ci/plugins/pipeline-maven/2.5.0-alpha-1/pipeline-maven-2.5.0-alpha-1.hpi

**Comment by justinbarry** [comment]
Thank you for bringing this home. I'm testing the artifact above now.

**Comment by cyrille-leclerc** [comment]
released as 2.5.0

**Comment by justinbarry** [comment]
Awesome. Thanks!

**Comment by cyrille-leclerc** [comment]
@justinbarry thanks. Can you please confirm that it works as desired? That you can see links on snapshot artifacts generated by a build and used as dependencies by other builds?

**Comment by justinbarry** [comment]
Amazingly, it was working before but some artifacts were being attributed to an incorrect original owner downstream. I'll need to let this work for a few days to see if we have remedied the problem.

Thank you for your work and I'll let you know if anything begins appearing incorrectly.

**Pull request closed without merging**